bson/value_reader.go

```diff
@@ -57,14 +72,26 @@ func newValueReader(t Type, r io.Reader) ValueReader {
 }
 
 func newDocumentReader(r io.Reader) *valueReader {
-	stack := make([]vrState, 1, 5)
-	stack[0] = vrState{
-		mode: mTopLevel,
-	}
-	return &valueReader{
-		r:     bufio.NewReader(r),
-		stack: stack,
-	}
+	vr := vrPool.Get().(*valueReader)
+
+	vr.offset = 0
+	vr.frame = 0
+
+	vr.stack = vr.stack[:1]
+	vr.stack[0].mode = mTopLevel
+
+	br := bufioReaderPool.Get().(*bufio.Reader)
+	br.Reset(r)
+	vr.r = br
+
+	return vr
+}
+
+func releaseDocumentReader(vr *valueReader) {
```

Comment:
Will moving the cleanup here from `newDocumentReader()` make the memory usage of the pool smaller?

Reply:
Are you suggesting we reset the valueReader fields in `releaseDocumentReader`? Decoders require these fields to be defensively reset when getting a new document read, otherwise you will get errors like this:

```
error unmarshaling BSON: ReadDocument can only read a Document while positioned on a TopLevel, Element, or Value but is positioned on a Unknown
```